### DISCUSSION TITLE
Shared (abstract) EEG/neuroimaging types

### DISCUSSION THREAD

**Discussion by marcpabst** [unfoldtoolbox/Unfold.jl#182] [General]
Hey all,

Not sure if it would be useful at all, but I'm starting to do more and more EEG analysis in Julia and I was wondering if there is a case for introducing some sort of shared (abstract) EEG and/or other neuroimaging types. They could behave like normal vectors or matrices but could also store EEG metadata and allow for time-based indexing. They could also easily be converted to and from other structures (e.g., to interface with `PyMNE`). 

I know that this is not necessarily a great fit for `Unfold.jl`, but seeing that you're already putting a lot of infrastructure for plotting and statistical analysis in place, I feel it's a good place to ask if there i any interest.

**Comment by behinger** [comment]
Hey!
Thanks for reaching out! I have seen this discussion multiple times, most recently I participated in NeuroCore.jl - but this is already a year ago or so...

Prior datastructures are defined in at least 
- Neuroimaging.jl (previously EEG.jl) (not active)
- NeuroAnalyzer.jl (active, 1 contributor, very large dependencies, https://discourse.julialang.org/t/ann-neuroanalyzer-jl/89574)
- NeuroAnalysis.jl (new, LFP)
- BrainWave.jl (deprecated)
- NeuroCore.jl (status unclear, https://github.com/JuliaNeuroscience/NeuroCore.jl/ - some discussion on dataformat in here)

I think NeuroAnalyzer.jl has most features, Neuroimaging.jl is the oldest, NeuroCore.jl in principle the right place, but no development even after asking from multiple people.

I'm interested in participating and I could be a long-term maintainer, but I dont have capacity to kickstat it. Does that make sense?

**Comment by marcpabst** [comment]
Hey Benedikt, thanks for getting back to me!

I actually remember that this topic came up a few times, and I participated in at least one of those discussions. Usually, I then stopped doing EEG for some time, and upon returning, noticed not much progress had been made. However, I think something has changed now - Julia is at a point where it is actually usable and ahead of the alternatives for a lot of numerical stuff!

In regards to the projects you mentioned, I tend to agree:
- `Neuroimaging.jl` looks nice but hasn't been updated in a long time.
- I'm not terribly excited about the approach `NeuroAnalyzer.jl` is taking, mostly for the reasons you mentioned.
- `NeuroCore.jl` indeed seems like the right approach, although I haven't seen it being used anywhere. Honestly, going through the code and discussions, it looks like `NeuroCore.jl` failed because it tried to do too much at once.. For example, I don't know of anyone who is doing fMRI stuff in Julia, but there seem to be a fair few people working with EEG/MEG/physiological data.

So my preliminary idea would be to

1. Start with a very lightweight set of abstract data types or traits. This could be based on Julia's `AbstractArray` types, or follow the `NeuroCore` approach - although I'm not sure traits are the best way to handle this (traits are great, but my understanding is that they lack first-class language support and don't integrate well with other parts of the Julia ecosystem. For example, I'm not sure if you can create Makie recipes based on traits? Also, they tend to be ugly in Julia).
2. Provide some useful infrastructure around these types, mainly by implementing a wrapper around `PyMNE` objects, so you get access to a lot of pre-existing infrastructure; and by providing `Makie` integration (I think `Unfold.jl` plotting tools are a really good place to start). For example, I'm yet to see a nice and repsonsive (!!) viewer for raw EEG data... 
3. Ping everyone who is working on these kinds of things in Julia and open PRs in projects to add support for these new types/traits.

Does this approach sound reasonable to you? Are there any other people you know of who might be interested in this project?

**Reply by behinger** [reply]
- I'm currently thinking to switching Unfold to traits for extensibility, but I totally agree with your worries. 
- it would be helpful to have a semi-systematic delineation of data-dimensionality (e.g. `sensor`,`time`,`epoch`,`frequency`) and their implementation (e.g. `sensor`x`sensor`x `time` is possible as well but not on everyones radar). I think AxisIndices.jl was put in the ring at some point?
- I'm open to move out all UnfoldMakie.jl generic code to remove the Unfold-dependency at some point

I'm annoyed wth PyMNE as well, given you have to do a lot of pyconvert after the (justified) switch from Pycall to Pythoncall. So starting a wrapper there sounds like a good place

Connecting to the people participating in the discussion and few others is probably a good idea. I can ask in my lab if anyone is interested to help out as well.

**Reply by marcpabst** [reply]
I think labeling dimensions is a good idea and `AxisIndices.jl` looks interesting, although no updates in 3 years unfortunately... Maybe https://github.com/JuliaArrays/AxisArrays.jl would be helpful?

I'm generally a bit wary of "abusing" language features to implement this kind of fundamental behavior (how many OOP systems does R have at the moment?) but traits seem to have become the generally accepted way of dealing with one-to-many relationships in Julia, so I would be okay with using them as long as we somehow enforce `AbstractArray`-ness of the data, so that all methods that accept arrays can still use them.

On the other hand, I'm also a big fan of composition-over-inheritance, so having to create you own `MyType :< AbstractEEGTimeseries` that wraps whatever you want to use doesn't seem like a bad choice to me.

Re `PyMNE`, I think a wrapper might even be a good start for a fully-fledged EEG package - one could start by thinly wrapping MNE functions and then step-by-step reimplement them for better performance.

**Reply by behinger** [reply]
Yeah AxisArrays.jl is even better! 

what would be the interface for an `AbstractEEGTimeseries`? Is it realistic to implement those functions yourself if you want your own type?

What concrete types should we offer if any?

just spitballing some simple getter / setters:
```julia
struct EEGTimeseries <: AbstractEEGTimeseries
     data = pyconvert(AxisArray,raw.get_data(unit="µV"))µV # with unitful
     raw = pyMNE.raw
end

sampling_rate(T)
T[1:5,2:3]
T[channel=["Fz","Oz"],time=0.1:5,epoch=1:5,source_idx=15]
channel_names(T)
events(T)
T.info, T.ch_names() # etc. => wrapper to the mne-raw object
```

something like this? We should also think about a good ordering of the data, e.g. eeglab is `ch x time x epoch` which I use in Unfold, but it is sometimes unfortunate, as the fastest dimension is not the first one. `time x ch x epoch` might be better, or `epoch x time x ch` (sorted by typical size)

**Reply by marcpabst** [reply]
I think you're raising important points here. I wonder if things should be more agnostic to the actual storage format and then automatically provide more consistent ways to access the data in a standardized way (i.e. with standadised ordering and maybe even normalised units?).

Especially because it would be great to have support for 'exotic' stuff like ragged arrays (for example, I sometimes work with data where epochs/segments don't have the exact same length).

Maybe traits are a good idea after all? I have worked a lot in Rust lately, whre traits a first-class citizens. I haven't really worked with Julia traits much, but I would probably define a `MultivariateTimeseriesTrait` trait or interface, that comes with
- a `data()` method, that returns the data as an `AbstractArray` - this can either be the real struct that's used to store the data or a view,
- an `axistypes()` method, that would map the axes of the array returned by `data()` to their corresponding types. E.g., it could return `[:sensor, :epoch, :time]`,
- an `axisunits()` method that returns the units of the corresponding axes (maybe as `Unitful` types?).

In theory, this should be enough to then automatically provide a time-and sensor-indexable view with well-defined dimension ordering. This could be done via another `ArrayTrait` - there are a few trait macros out there that all work to various extents, but ideally you could then just (also just spitballing here and completely untested):

```julia
using SimpleTraits

@traitdef MultivariateTimeseriesTrait{X}
@traitfn data(x::X) where {X; MultivariateTimeseriesTrait{X}} = @error "data method must be implemented for $(X)"
@traitfn axistypes(x::X) where {X; MultivariateTimeseriesTrait{X}} = @error "axistypes method must be implemented for $(X)"
@traitfn axisunits(x::X) where {X; MultivariateTimeseriesTrait{X}} = @error "axisunits method must be implemented for $(X)"

@traitdef ArrayTrait{X}
@traitfn Base.eltype(x::X) where {X; ArrayTrait{X}} = eltype(data(x))
@traitfn Base.length(x::X) where {X; ArrayTrait{X}} = length(data(x))
@traitfn Base.size(x::X) where {X; ArrayTrait{X}} = size(data(x))
# ... other array methods + some magic for indexing using sensor names, times, and so on
# this could also be used to implement support for Unitful
```

And then for a package that consume the core package:
```julia
using AbstractNeuroTimeseries # or whatever
using SimpleTraits

# define a EEGTimeseries type
struct EEGTimeseries{T,N} <: AbstractArray{T,N}
    raw = PyMNE.Py
end

# implement both traits for EEGTimeseries
@traitimpl MultivariateTimeseriesTrait{EEGTimeseries}
@traitimpl ArrayTrait{EEGTimeseries}

# overwrite necessary methods for EEGTimeseries
data(x::EEGTimeseries) = PyMNE.pyconvert(PyMNE.PyArray, x.raw._data)
axistypes(x::EEGTimeseries) = [:sensors, :time]
axisunits(x::EEGTimeseries) = [:V, :s]
```

- I just wish traits had the same sort of support as they have in Rust - would make things so much easier...
- I just realsied that the code above won't work the way I hoped because `AbstractArray` already defines some methods that are more specific thant what  `SimpleTraits` generates - so maybe an`AbstractEEGTimeseries :< AbstractArray` would be necessary after all
- the above code would be extremely slow, because eery indexing oepration would require a call into Python (but this can easily be fixed)

**Reply by behinger** [reply]
- Interesting concept, haven't seen a package like this so far. It feels like an interface-package; cool! Are there other CorePackages that use this approach? Maybe some lessons learned from them?
- I think an `AbstractEEGTimeseries` is completely expected. Probably even easier for most people to use (like me, no ideas about traits ;-))
- RaggedArrays are useful indeed, i think only fieldtrip really supports them right now, so would be helpful to have.

I think we could do the following:

1. Make a list of people that could be interested (don't ping yet)
2. Make an initial collection of functions / accessors / the interface, which we can use as a first proposition to the people on the top list. We just have to be careful not to make too strong decisions in the design, we don't want to be exclusive. But I do want to have something already for discussion, because earlier efforts died down realtively quickly. 
3. We would need this at a place where it is easy to provide comments to individual functions (e.g. `I don't like to call it axisunits, it should just be units` - that kind of discussion). Google docs feels clumsy. Maybe a github-discussion could actually work?

(4. Profit!)

**Reply by marcpabst** [reply]
I tend to agree that providing an `AbstractEEGTimeseries` type would be the easiest and least confusing solution. I even think the same approach as I have outlined above would work—a core package could define all the necessary methods to make fancy indexing stuff work, and a downstream library would just need to provide the data and information about the axes and units (e.g., by implementing `data()`, `units()`, and `axes()` methods).

I also think your suggested approach is reasonable - a few notes:
 - regarding people, a few that come to mind (plus a few a quick Google search turned up): wizofe, leoscholl, rob-luke, sam81m, wherrera10, tokazama, azraq27, likanzhan, and VMHidalgo, anyone at Beacon Biosignals. I have some hope that keeping the discussion on GitHub would lower the barrier for people to actually comment....
- I have a working implementation based on code that I personally use (slightly cleaned up, but not extensvly tested or anyhting) [here](https://github.com/marcpabst/NeuroTimeseries.jl/blob/master/test.jl)
 - I think the key is to keep things a) lightweight and very easy to implement, and b) to provide some initial infrastructure
- I like profit!

**Reply by behinger** [reply]
how did you implement this so fast & clean - impressed!

- Minor: `sensor & channels` - I never grasped when which is used. My feeling, sensor => MEG-dominant, channel => EEG-dominant
- how to handle metadata / annotation / events (please let us never repeat the MNE mistake to have 3 to 4 different ways to represent effectively the same thing...)? `events(X)::AbstractTable` ? or can we even define a second interface for the events to define e.g. `onsets(E) #in seconds`, `latencies(E) # in samples`
- dimensionality: diff between logical and actual is unclear to me, if you have 1ch, why not simply store as vector internally?

people: - cbrnr, palday (beacon) for sure (both MNE developers), jschepers from my lab could be interested 


I didnt really get why some functions have a `_` in front in your explanation, I'm missing something here - is this SimpleTraits magic?

**Reply by marcpabst** [reply]
>how did you implement this so fast & clean - impressed!

Haha, I've been using something similar for a while, so it was mostly a matte rof cleaning things up a little.

> Minor: sensor & channels - I never grasped when which is used. My feeling, sensor => MEG-dominant, channel => EEG-dominant

I think this is a classical case of different traditions stemming from different lines of research. I personally prefer channels because it feels closer to actually describing data instead of the data collection (and are passive EEG electrodes even considered sensors?). But then again, I've never worked with MEG and I guess it makes sense to call them "sensors" once you talk about locations of physical MEG sensors. But honestly, I don't mind calling them either...

> how to handle metadata / annotation / events (please let us never repeat the MNE mistake to have 3 to 4 different ways to represent effectively the same thing...)? events(X)::AbstractTable ? or can we even define a second interface for the events to define e.g. onsets(E) # in seconds, latencies(E) # in samples

Totally agree! How about borrowing from the BIDS specs here. We could have an `AbstractEEGMetadata <: AbstractDict{Symbol,Any}` type that would contain information as described in https://bids-specification.readthedocs.io/en/stable/modality-specific-files/electroencephalography.html. Also, it would be possible to automatically (via a macro) create accessor methods, so that you can call both `eegreference(metadata)` and `metadata[:eegreference]`.

The same could be done for events/triggers - as you said, they could conform to `AbstractTable` (or `AbstractDataFrame`, no idea what's better) so that it's very easy to extract data and manipulate events. On top of that, we could have the other interface you described, i.e. `onsets()`/`latencies()`, `duration()` and maybe even all the other BIDS-defined `events.tsv` columns like `trialtype()`, `responsetime()`, etc. 

The most obvious benefit of this would be that converting from/to BIDS files would be extremely trivial and we could potentially leverage `mne_bids` to easily convert from and to MNE info objects. Also, just blindly following the BIDS specs would save a lot of time!

**Reply by behinger** [reply]
- I like the BIDS interface and following it!

- Thus we would have `AbstractEEGTimeseries`, `AbstractEEGMetadata` & `AbstractEvents`. The latter two should implement at least the mandatory fields from the BIDS datastructure maybe? 
- I mean in principle we could 1:1 copy the BIDS interface and implement all "files" as one AbstractType? E.g. `AbstractChannels`, `AbstractEvents`, `AbstractElectrodes`, `AbstractCoordsystem`, `AbstractEEG` (what you called `AbstractMetadata` - I actually prefer your name there, `AbstractEEG` should be the container containing all the others maybe, )*

- `Sensor` vs. `Channels`: Even BIDS seems to use them interchangeable, so implement both. `electrodes` is used differently though


\* I noticed I already dropped the `EEG`, because why not ;)

**Reply by marcpabst** [reply]
> I mean in principle we could 1:1 copy the BIDS interface and implement all "files" as one AbstractType? E.g. AbstractChannels, AbstractEvents, AbstractElectrodes, AbstractCoordsystem, AbstractEEG (what you called AbstractMetadata - I actually prefer your name there, AbstractEEG should be the container containing all the others maybe, )*

I like that! I did some pondering, and I think we could even split the interface into both types and traits, so that you don't need to use the provided types but can still resort to them for easier implementation. Btw, I think Tables.jl is a good example of leveraging traits for this.

- For example, the `events()` methods called on a EEG timeseries would be required to return a type that implements the `isevents` trait (providing the `onsets()` and `durations()` methods, which could be part of the default implementation of the `AbstractEvents` type (which could also implement the `Tables.istable()` trait).
- Same for `electrodes()`, which would need to return a type implementing `iselectrodes`, which ensures compatibility with the BIDS standard for encoding electrode information/location. Default implementation would be provided through the `AbstractEEGElectrodes` type (which again could implement the Tables.jl interface by default)
- As you said, the default type for `metadata()` could be called `AbstractMetadata` (or maybe better `AbstractEEGMetadata`) which would then implement the `iseegmetadata()` trait, and in turn either provide all the allowed BIDS fields as methods (and by default just return `nothing`) or make them accessible through a `getfield(m, ::Symbol)` method.

Using traits instead of types would allow to implement these things on top of existing `AbstractTypes` - so you might decide to inherit from an `AbstractArray` or `AbstractVector` for your events instead of the provided `AbstractEvents` type.

At the top of the (logical) hierarchy, we could then have an `AbstractMultivariateTimeseries` type (and we also could have `AbstractEEG` and `AbstractMEG` based on that) - this type in turn implements different useful traits using default/fallback methods, like the `istimeseries` (see below) trait. On top of that, this could also implemen the `Tables.istable` trait, so it can be fed to anything that accepts a Tables.jl table, like CSV.jl, AlgebraOfGraphic - you can even do `DataFrame(ts::AbstractMultivariateTimeseries)`. `AbstractMultivariateTimeseries` could then also provide the `AbstractArray` interface (if we want that).

And finally, the `istimeseries` trait would provide standardised access to the underlying data using a `data()`method and also come with a set of other useful methods like `duration()`, `nsamples`, `nobservations`, and so on. Plus, it could define methods to get metadata through: `metadata()`, `events()`, and `observations()` as described above.

**Reply by behinger** [reply]
I tried to understand Tables.jl, but the explanations are relatively tricky because quite abstract.

- Regarding `AbstractMetadata` vs `AbstractEEGMetadata` - I was thinking that BIDS uses the same metadata everywhere. could be `AbstractEEGMetadata <: AbstractMetadata` hierarchy, but let's focus on the EEG bit first :-D
- what are `observations` in your terminology? repetitions/trials?

**Reply by marcpabst** [reply]
I agree, Tables.jl is not super straightforward... Here's my undertanding of how it works: When you want to accept a Table-like type `t::SomeType`, you check if `Tables.istable(t)` is `true` (this can be done at compile time, Holy-trait-style). If yes, we say `t` implements the trait `istable`. 

You then check if `t` allows for column-access (or row-access) by calling `Tables.allowscolumnaccess(t)`. If `true`, you can use `c = Tables.columns(t)` to get a type that imlements the `AbstractColumns` interface, i.e., `Tables.getcolumn(c, i::Int)`, `getcolumn(c, nm::Symbol)`, `columnnames(c)`  (**and might also subtype `AbstractColumns`, but doesnt have to**). 

And because it's all implemented via traits, you can have `Tables.columns(t) = t`, if your original type implements the columns interface.

> Regarding AbstractMetadata vs AbstractEEGMetadata - I was thinking that BIDS uses the same metadata everywhere. could be AbstractEEGMetadata <: AbstractMetadata hierarchy, but let's focus on the EEG bit first :-D

I think keeping it as non-specific as reasonably possible is a really good idea!

> what are observations in your terminology? repetitions/trials?

I mainly tried to avoid "epochs', because it feels so EEG speicfic. "Observation" feels more in line with the general ML lingo (not sure what the "traditional stats" term would be - "repetition"?) Maybe a EEG subtype could use "epochs" instead?

I have some updated code (and a simple notebook with examples) I'll try to share shortl for some feedback (currently using real, unpublished data, so I will need to replace that with some mock data) :)

**Reply by behinger** [reply]
Hey! Anything you can share? 
There is now also https://slopezpereyra.github.io/EEGToolkit.jl/ - a new package 

Happy to spend some time working on this :)

**Reply by behinger** [reply]
@marcpabst bump?
